05-Mining-Protocol.md

```diff
@@ -1,101 +1,182 @@
+
+So the protocol establishes the notion of **Extended Extranonce**, which is a fundamental component of Extended Jobs.
 
-Whenever a standard channel is created it is always put into some channel group identified by its `group_channel_id`.
-Group channel ID namespace is the same as channel ID namespace on a particular connection but the values chosen for group channel IDs must be distinct.
+The Extended Extranonce is an array of 32 bytes, split into three different areas:
 
-### 5.1.4 Future Jobs
+![](./img/extended_extranonce.png)
 
-An empty future block job or speculated non-empty job can be sent in advance to speedup new mining job distribution.
-The point is that the mining server MAY have precomputed such a job and is able to pre-distribute it for all active channels.
-The only missing information to start to mine on the new block is the new prevhash.
+- The `extranonce_prefix` bytes are reserved for the upstream layer, where fixed bytes were already established for the Extranonce and are no longer available for rolling or search space splitting.
+- The **locally reserved** bytes is where the proxy will assign unique values for its own downstream clients, according to their hashpower.
+- The **downstream reserved** bytes is where the downstream clients will further distribute the search space (if they are also proxies), or use for rolling (if they are SV1 mining devices, or SV2 mining devices over 280 TH/s).
+
+![](./img/extended_extranonce_proxies.png)
+
+In order to calculate the Merkle Root, an Extended Job carries the following data:
+- `merkle_path`
+- `coinbase_prefix`
+- `coinbase_suffix`
+
+An [Extended Channel](#532-extended-channels) has the following properties:
+- `extranonce_prefix`: the Extended Extranonce bytes that were already allocated by the upstream server.
+- `extranonce_size`: how many bytes are available for the locally reserved and downstream reserved areas of the Extended Extranonce.
+
+And a [Standard Channel](#531-standard-channels) has the following property:
+- `extranonce_prefix` the Extended Extranonce bytes that were already allocated by the upstream server.
+
+So when a proxy receives an Extended Job, it could either:
+- propagate it as an Extended Job to a downstream Extended or Group Channel, where the Extended Extranonce will be further split.
+- convert it into multiple Standard Jobs, where each downstream Standard Channel's `extranonce_prefix` is combined with `coinbase_prefix` + `coinbase_suffix` to calculate a unique Merkle Root.
+
+### 5.1.3 Future Jobs
+
+A Job with an empty template or speculated non-empty template can be sent in advance to speedup Job distribution when a new block is found on the network.
+
+The mining server MAY have precomputed such a Job and is able to pre-distribute it for all active Channels.
+The only missing information to start to mine on the new block is the new `prev_hash`.
```

Comment:
on [his review](https://github.com/plebhash/sv2-spec/commit/e874f121423a2667e8731cdd41a5922697f5578d) @mcelrath said:

> BOB: And timestamp, and Braidpool metadata -- more fine-grained mallation of the Merkle Tree and coinbase will be needed for Braidpool.

@mcelrath can you elaborate? are you saying that the spec will need to be adapted in order to cater for Braidpool?

please keep in mind that this section is about Future Jobs, and I don't really understand if your comment is about Future Jobs only, or on a more broad sense

Reply:
What I'm saying is that when a block (or share) is found, within Braidpool we have to update more than just `prev_hash`. We also have to update DAG parents, and possibly our metadata. If we assume all that is encoded in some TX somewhere in the Merkle tree (e.g. an OP_RETURN or pubhash tweak) we would need to update that as well.

Reply:
So, can we create a slightly more general mechanism that allows additional malleation of the coinbase tx? I think providing a new `prev_hash` and full coinbase tx would do it.

Reply:
as stated in the PR description, the goal of this PR is not to change the spec in any way.

the goal here is simply to make the specs for the Mining subprotocol more clear for the reader to understand.

a viable path forward is proposing a protocol extension.

I started this discussion so we can move forward with the idea: https://github.com/stratum-mining/sv2-spec/discussions/100